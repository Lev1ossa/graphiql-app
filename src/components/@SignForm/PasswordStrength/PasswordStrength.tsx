import { useLocale } from '@/hooks/useLocale'
import { useValidators } from '@/hooks/useValidators'
import { Box, FormHelperText } from '@mui/material'
import { Fragment } from 'react'
import { ZodError, type ZodString } from 'zod'
import PasswordStrengthBar from './PasswordStrengthBar'

const getPasswordErrors = (password: string, schema: ZodString) => {
  try {
    schema.parse(password)
  } catch (e) {
    if (e instanceof ZodError) {
      return e.issues.map((issue) => issue.message)
    }
  }
  return []
}

export default function PasswordStrength(props: {
  password: string
  disabled: boolean
}) {
  const { password = '', disabled } = props
  const {
    password: { schema, strength, errors: validatorErrors },
  } = useValidators()
  const {
    locale: { signInUpPage },
  } = useLocale()

  const passwordErrors = getPasswordErrors(password, schema)

  const shouldContain = () => {
    return (
      <Box component="span">
        {validatorErrors.map(({ long, short }, index) => {
          const color = passwordErrors[0] === long ? 'error.main' : 'inherit'
          return (
            <Fragment key={short}>
              <Box component="span" color={color}>
                {short}
              </Box>
              {index !== validatorErrors.length - 1 ? ', ' : ''}
            </Fragment>
          )
        })}
      </Box>
    )
  }

  const currentStrength = strength.length - passwordErrors.length - 1

  return (
    <Box px={1}>
      <FormHelperText>
        {signInUpPage.typography.passwordStrength.description} {shouldContain()}
        <PasswordStrengthBar
          disabled={disabled}
          current={currentStrength}
          levels={strength.map(([color, text]) => ({ color, text }))}
        />
      </FormHelperText>
    </Box>
  )
}
