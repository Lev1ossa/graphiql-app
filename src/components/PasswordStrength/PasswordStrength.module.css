.inner_indicator {
  height: 5px;
  width: 50px;
  background-color: grey;
  display: inline-block;
  margin-right: 10px;
  border-radius: 2px;
  transition: background-color 0.2s;
}

.red .inner_indicator:first-child {
  background-color: #d32f2f;
  transition: background-color 0.2s;
}

.orange .inner_indicator:first-child {
  background-color: #ef6c00;
  transition: background-color 0.2s;
}

.orange .inner_indicator:nth-child(2) {
  background-color: #ef6c00;
  transition: background-color 0.2s;
}

.yellow .inner_indicator:first-child {
  background-color: yellow;
  transition: background-color 0.2s;
}

.yellow .inner_indicator:nth-child(2) {
  background-color: yellow;
  transition: background-color 0.2s;
}

.yellow .inner_indicator:nth-child(3) {
  background-color: yellow;
  transition: background-color 0.2s;
}

.salad .inner_indicator:first-child {
  background-color: #98bf54;
  transition: background-color 0.2s;
}

.salad .inner_indicator:nth-child(2) {
  background-color: #98bf54;
  transition: background-color 0.2s;
}

.salad .inner_indicator:nth-child(3) {
  background-color: #98bf54;
  transition: background-color 0.2s;
}

.salad .inner_indicator:nth-child(4) {
  background-color: #98bf54;
  transition: background-color 0.2s;
}

.green .inner_indicator {
  background-color: #2e7d32;
  transition: background-color 0.2s;
}
