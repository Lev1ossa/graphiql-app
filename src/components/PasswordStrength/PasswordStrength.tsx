import { usePasswordErrors } from '@/hooks/usePasswordErrors'
import { type userSignInData, type userSignUpData } from '@/types/types'
import { getPasswordErrorsData } from '@/utils/getPasswordErrorsData'
import { Box, FormHelperText, Typography } from '@mui/material'
import { type UseFormGetValues } from 'react-hook-form'
import styles from './PasswordStrength.module.css'

export function PasswordStrength(props: {
  watch: UseFormGetValues<userSignUpData | userSignInData>
}) {
  const { watch } = props

  const passwordErrors = usePasswordErrors(watch('password'))

  return (
    <>
      {passwordErrors && (
        <Box
          sx={{
            minWidth: '552px',
            position: 'absolute',
            top: '60px',
            display: 'flex',
            paddingLeft: '15px',
          }}
        >
          <div
            className={getPasswordErrorsData(passwordErrors, styles).className}
          >
            <span className={styles.inner_indicator} />
            <span className={styles.inner_indicator} />
            <span className={styles.inner_indicator} />
            <span className={styles.inner_indicator} />
            <span className={styles.inner_indicator} />
          </div>
          <Typography variant="body2">
            {getPasswordErrorsData(passwordErrors, styles).text}
          </Typography>
        </Box>
      )}
      {passwordErrors && passwordErrors.length > 0 && (
        <FormHelperText
          sx={{ minWidth: '542px', position: 'absolute', top: '80px' }}
          error
        >
          {`Strong password must contain ${passwordErrors?.join(', ')}`}
        </FormHelperText>
      )}
    </>
  )
}
