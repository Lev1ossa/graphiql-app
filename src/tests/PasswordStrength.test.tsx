import { PasswordStrength } from '@/components/@SignForm/PasswordStrength'
import { expect, test } from 'vitest'
import renderInContext from './renderInContext'

test('Correct PasswordStrength snapshot', () => {
  const { baseElement } = renderInContext(
    <PasswordStrength password="" disabled={false} />,
  )
  expect(baseElement).toMatchSnapshot()
})
