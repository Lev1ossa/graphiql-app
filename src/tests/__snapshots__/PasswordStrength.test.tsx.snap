// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Correct PasswordStrength snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-mro3c9"
    >
      <p
        class="MuiFormHelperText-root css-1d1r5q-MuiFormHelperText-root"
      >
        Strong password contains at least
         
        <span
          class="MuiBox-root css-0"
        >
          <span
            class="MuiBox-root css-lhj5ms"
          >
            8 characters
          </span>
          , 
          <span
            class="MuiBox-root css-q2y3yl"
          >
            uppercase letter
          </span>
          , 
          <span
            class="MuiBox-root css-q2y3yl"
          >
            lowercase letter
          </span>
          , 
          <span
            class="MuiBox-root css-q2y3yl"
          >
            one digit
          </span>
          , 
          <span
            class="MuiBox-root css-q2y3yl"
          >
            one special character
          </span>
        </span>
        <span
          class="MuiStack-root css-1g7v78n-MuiStack-root"
        >
          <span
            class="MuiBox-root css-1dqsri6"
          />
          <span
            class="MuiBox-root css-1dqsri6"
          />
          <span
            class="MuiBox-root css-1dqsri6"
          />
          <span
            class="MuiBox-root css-1dqsri6"
          />
          <span
            class="MuiBox-root css-1dqsri6"
          />
          <span
            class="MuiBox-root css-1f5qj6m"
          />
        </span>
      </p>
    </div>
  </div>
</body>
`;
